File: youtube_service.py
from flask import Flask, request, jsonify
from flask_cors import CORS
import yt_dlp
import requests
import logging
import time

app = Flask(__name__)
CORS(app)

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Invidious 实例列表
INVIDIOUS_INSTANCES = [
    "https://inv.nadeko.net",
    "https://invidious.jing.rocks",
    "https://invidious.privacyredirect.com",
    "https://y.com.sb",
]

@app.route('/')
def home():
    return '''YouTube 音乐提取服务 (yt-dlp + Invidious)

使用方法:
1. 搜索: /search?q=关键词
2. 获取音频: /audio?id=视频ID

优先使用 yt-dlp，失败时自动切换到 Invidious
'''

@app.route('/search')
def search():
    query = request.args.get('q')
    if not query:
        return jsonify({'error': '缺少搜索关键词'}), 400

    logger.info(f'搜索请求: {query}')

    # 优先使用 yt-dlp
    try:
        results = search_ytdlp(query)
        if results:
            logger.info(f'yt-dlp 搜索成功，返回 {len(results)} 条结果')
            return jsonify(results)
    except Exception as e:
        logger.warning(f'yt-dlp 搜索失败: {str(e)}，尝试 Invidious')

    # 回退到 Invidious
    try:
        results = search_invidious(query)
        logger.info(f'Invidious 搜索成功，返回 {len(results)} 条结果')
        return jsonify(results)
    except Exception as e:
        logger.error(f'所有搜索方法都失败了: {str(e)}')
        return jsonify({'error': f'搜索失败: {str(e)}'}), 500

@app.route('/audio')
def audio():
    video_id = request.args.get('id')
    video_url = request.args.get('url')

    if not video_id and not video_url:
        return jsonify({'error': '缺少视频ID或URL'}), 400

    if video_url:
        video_id = extract_video_id(video_url)
        if not video_id:
            return jsonify({'error': '无效的YouTube URL'}), 400

    logger.info(f'获取音频: {video_id}')

    # 优先使用 yt-dlp
    try:
        result = get_audio_ytdlp(video_id)
        if result:
            logger.info(f'yt-dlp 获取成功: {result["title"]}')
            return jsonify(result)
    except Exception as e:
        logger.warning(f'yt-dlp 获取失败: {str(e)}，尝试 Invidious')

    # 回退到 Invidious
    try:
        result = get_audio_invidious(video_id)
        logger.info(f'Invidious 获取成功: {result["title"]}')
        return jsonify(result)
    except Exception as e:
        logger.error(f'所有获取方法都失败了: {str(e)}')
        return jsonify({'error': f'获取音频失败: {str(e)}'}), 500

# ========== yt-dlp 实现 ==========

def search_ytdlp(query):
    ydl_opts = {
        'quiet': True,
        'no_warnings': True,
        'extract_flat': True,
        'default_search': 'ytsearch10',
    }

    with yt_dlp.YoutubeDL(ydl_opts) as ydl:
        result = ydl.extract_info(f'ytsearch10:{query}', download=False)

        if not result or 'entries' not in result:
            return []

        results = []
        for entry in result['entries']:
            if not entry:
                continue

            duration = format_duration(entry.get('duration', 0))

            results.append({
                'id': entry.get('id', ''),
                'title': entry.get('title', ''),
                'author': entry.get('uploader', ''),
                'duration': duration,
                'thumbnail': entry.get('thumbnail', '')
            })

        return results

def get_audio_ytdlp(video_id):
    ydl_opts = {
        'quiet': True,
        'no_warnings': True,
        # 优先选择最低画质的视频（包含音频），如果没有则选择纯音频
        'format': 'worst[height<=360]/bestaudio/best',
        'extract_flat': False,
        # 添加更多选项以绕过 bot 检测
        'extractor_args': {'youtube': {'player_client': ['android', 'web']}},
        'user_agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
    }

    with yt_dlp.YoutubeDL(ydl_opts) as ydl:
        info = ydl.extract_info(f'https://www.youtube.com/watch?v={video_id}', download=False)

        # 获取音频/视频 URL
        audio_url = None
        if 'url' in info:
            audio_url = info['url']
        elif 'formats' in info:
            # 优先选择有音频的最低画质视频
            for fmt in info['formats']:
                if fmt.get('acodec') != 'none':
                    audio_url = fmt.get('url')
                    break

            if not audio_url and info['formats']:
                audio_url = info['formats'][-1].get('url')

        if not audio_url:
            raise Exception('没有找到音频流')

        return {
            'url': audio_url,
            'title': info.get('title', ''),
            'author': info.get('uploader', ''),
            'duration': format_duration_long(info.get('duration', 0))
        }

# ========== Invidious 实现 ==========

def search_invidious(query):
    for instance in INVIDIOUS_INSTANCES:
        try:
            url = f"{instance}/api/v1/search?q={requests.utils.quote(query)}&type=video"
            resp = requests.get(url, timeout=10)

            if resp.status_code != 200:
                continue

            data = resp.json()
            results = []

            for item in data:
                if item.get('type') != 'video':
                    continue

                duration = format_duration(item.get('lengthSeconds', 0))
                thumbnail = ''
                if item.get('videoThumbnails'):
                    thumbnail = item['videoThumbnails'][0].get('url', '')

                results.append({
                    'id': item.get('videoId', ''),
                    'title': item.get('title', ''),
                    'author': item.get('author', ''),
                    'duration': duration,
                    'thumbnail': thumbnail
                })

                if len(results) >= 10:
                    break

            return results

        except Exception as e:
            logger.debug(f'Invidious 实例 {instance} 失败: {str(e)}')
            continue

    raise Exception('所有 Invidious 实例都失败了')

def get_audio_invidious(video_id):
    for instance in INVIDIOUS_INSTANCES:
        try:
            url = f"{instance}/api/v1/videos/{video_id}"
            resp = requests.get(url, timeout=10)

            if resp.status_code != 200:
                continue

            data = resp.json()

            # 选择最佳音频格式
            best_audio_url = None
            best_bitrate = 0

            for fmt in data.get('adaptiveFormats', []):
                if fmt.get('type', '').startswith('audio/'):
                    bitrate = int(fmt.get('bitrate', '0'))
                    if bitrate > best_bitrate:
                        best_bitrate = bitrate
                        best_audio_url = fmt.get('url')

            if not best_audio_url:
                continue

            return {
                'url': best_audio_url,
                'title': data.get('title', ''),
                'author': data.get('author', ''),
                'duration': format_duration_long(data.get('lengthSeconds', 0))
            }

        except Exception as e:
            logger.debug(f'Invidious 实例 {instance} 失败: {str(e)}')
            continue

    raise Exception('所有 Invidious 实例都失败了')

# ========== 工具函数 ==========

def format_duration(seconds):
    """格式化时长为 M:SS 或 H:MM:SS"""
    if not seconds:
        return "未知"

    # 确保是整数
    seconds = int(seconds)

    hours = seconds // 3600
    minutes = (seconds % 3600) // 60
    secs = seconds % 60

    if hours > 0:
        return f"{hours}:{minutes:02d}:{secs:02d}"
    return f"{minutes}:{secs:02d}"

def format_duration_long(seconds):
    """格式化时长为 XmYs"""
    if not seconds:
        return "未知"

    # 确保是整数
    seconds = int(seconds)

    hours = seconds // 3600
    minutes = (seconds % 3600) // 60
    secs = seconds % 60

    if hours > 0:
        return f"{hours}h{minutes}m{secs}s"
    return f"{minutes}m{secs}s"

def extract_video_id(url):
    """从 YouTube URL 提取视频 ID"""
    import re
    patterns = [
        r'(?:v=|\/)([0-9A-Za-z_-]{11}).*',
        r'(?:embed\/)([0-9A-Za-z_-]{11})',
        r'(?:watch\?v=)([0-9A-Za-z_-]{11})'
    ]

    for pattern in patterns:
        match = re.search(pattern, url)
        if match:
            return match.group(1)
    return None

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=8080)

File: youtube_service.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"net/url"
	"time"
)

type SearchResult struct {
	ID        string `json:"id"`
	Title     string `json:"title"`
	Author    string `json:"author"`
	Duration  string `json:"duration"`
	Thumbnail string `json:"thumbnail"`
}

type AudioInfo struct {
	URL      string `json:"url"`
	Title    string `json:"title"`
	Author   string `json:"author"`
	Duration string `json:"duration"`
}

func main() {
	http.HandleFunc("/search", handleSearch)
	http.HandleFunc("/audio", handleAudio)
	http.HandleFunc("/", handleRoot)

	port := "8080"
	log.Printf("YouTube 音乐服务启动在端口 %s", port)
	log.Printf("搜索: /search?q=关键词")
	log.Printf("获取音频: /audio?id=视频ID")
	log.Fatal(http.ListenAndServe(":"+port, nil))
}

func handleRoot(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "text/plain; charset=utf-8")
	fmt.Fprintf(w, "YouTube 音乐提取服务\n\n")
	fmt.Fprintf(w, "使用方法:\n")
	fmt.Fprintf(w, "1. 搜索: /search?q=关键词\n")
	fmt.Fprintf(w, "2. 获取音频: /audio?id=视频ID\n")
}

func handleSearch(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Access-Control-Allow-Origin", "*")
	w.Header().Set("Content-Type", "application/json; charset=utf-8")

	query := r.URL.Query().Get("q")
	if query == "" {
		http.Error(w, `{"error":"缺少搜索关键词"}`, http.StatusBadRequest)
		return
	}

	log.Printf("搜索请求: %s", query)

	results, err := searchViaInvidious(query)
	if err != nil {
		log.Printf("搜索失败: %v", err)
		http.Error(w, fmt.Sprintf(`{"error":"搜索失败: %s"}`, err.Error()), http.StatusInternalServerError)
		return
	}

	log.Printf("搜索成功，返回 %d 条结果", len(results))
	json.NewEncoder(w).Encode(results)
}

func handleAudio(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Access-Control-Allow-Origin", "*")
	w.Header().Set("Content-Type", "application/json; charset=utf-8")

	videoID := r.URL.Query().Get("id")
	videoURL := r.URL.Query().Get("url")

	if videoID == "" && videoURL == "" {
		http.Error(w, `{"error":"缺少视频ID或URL"}`, http.StatusBadRequest)
		return
	}

	if videoURL != "" {
		id, err := extractVideoID(videoURL)
		if err != nil {
			http.Error(w, fmt.Sprintf(`{"error":"无效的YouTube URL: %s"}`, err.Error()), http.StatusBadRequest)
			return
		}
		videoID = id
	}

	log.Printf("获取音频: %s", videoID)

	// 使用 Invidious API 获取音频
	audioInfo, err := getAudioViaInvidious(videoID)
	if err != nil {
		log.Printf("获取音频失败: %v", err)
		http.Error(w, fmt.Sprintf(`{"error":"获取音频失败: %s"}`, err.Error()), http.StatusInternalServerError)
		return
	}

	log.Printf("成功获取音频: %s - %s", audioInfo.Title, audioInfo.Author)
	json.NewEncoder(w).Encode(audioInfo)
}

// Invidious 搜索响应结构
type invidiousSearchResult struct {
	Type            string `json:"type"`
	Title           string `json:"title"`
	VideoID         string `json:"videoId"`
	Author          string `json:"author"`
	LengthSeconds   int    `json:"lengthSeconds"`
	VideoThumbnails []struct {
		URL string `json:"url"`
	} `json:"videoThumbnails"`
}

// Invidious 视频信息响应
type invidiousVideoInfo struct {
	Title           string `json:"title"`
	Author          string `json:"author"`
	LengthSeconds   int    `json:"lengthSeconds"`
	AdaptiveFormats []struct {
		URL      string `json:"url"`
		Type     string `json:"type"`
		Bitrate  string `json:"bitrate"`
		AudioQuality string `json:"audioQuality"`
	} `json:"adaptiveFormats"`
}

func searchViaInvidious(query string) ([]SearchResult, error) {
	instances := []string{
		"https://inv.nadeko.net",
		"https://invidious.jing.rocks",
		"https://invidious.privacyredirect.com",
		"https://y.com.sb",
	}

	var lastErr error
	for _, instance := range instances {
		results, err := tryInvidiousSearch(instance, query)
		if err == nil {
			return results, nil
		}
		lastErr = err
		log.Printf("搜索实例 %s 失败: %v", instance, err)
	}

	return nil, fmt.Errorf("所有实例都失败了: %v", lastErr)
}

func tryInvidiousSearch(instance, query string) ([]SearchResult, error) {
	searchURL := fmt.Sprintf("%s/api/v1/search?q=%s&type=video", instance, url.QueryEscape(query))

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Get(searchURL)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("HTTP %d", resp.StatusCode)
	}

	var invResults []invidiousSearchResult
	if err := json.NewDecoder(resp.Body).Decode(&invResults); err != nil {
		return nil, err
	}

	results := make([]SearchResult, 0, len(invResults))
	for _, item := range invResults {
		if item.Type != "video" {
			continue
		}

		duration := formatDuration(item.LengthSeconds)
		thumbnail := ""
		if len(item.VideoThumbnails) > 0 {
			thumbnail = item.VideoThumbnails[0].URL
		}

		results = append(results, SearchResult{
			ID:        item.VideoID,
			Title:     item.Title,
			Author:    item.Author,
			Duration:  duration,
			Thumbnail: thumbnail,
		})

		if len(results) >= 10 {
			break
		}
	}

	return results, nil
}

func getAudioViaInvidious(videoID string) (*AudioInfo, error) {
	instances := []string{
		"https://inv.nadeko.net",
		"https://invidious.jing.rocks",
		"https://invidious.privacyredirect.com",
		"https://y.com.sb",
	}

	var lastErr error
	for _, instance := range instances {
		info, err := tryInvidiousAudio(instance, videoID)
		if err == nil {
			return info, nil
		}
		lastErr = err
		log.Printf("音频实例 %s 失败: %v", instance, err)
	}

	return nil, fmt.Errorf("所有实例都失败了: %v", lastErr)
}

func tryInvidiousAudio(instance, videoID string) (*AudioInfo, error) {
	videoURL := fmt.Sprintf("%s/api/v1/videos/%s", instance, videoID)

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Get(videoURL)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("HTTP %d", resp.StatusCode)
	}

	var videoInfo invidiousVideoInfo
	if err := json.NewDecoder(resp.Body).Decode(&videoInfo); err != nil {
		return nil, err
	}

	// 选择最佳音频格式
	var bestAudioURL string
	var bestBitrate int

	for _, format := range videoInfo.AdaptiveFormats {
		// 只选择音频格式
		if format.Type != "" && (format.Type[:10] == "audio/mp4;" || format.Type[:11] == "audio/webm;") {
			// 尝试解析比特率
			var bitrate int
			fmt.Sscanf(format.Bitrate, "%d", &bitrate)

			if bitrate > bestBitrate {
				bestBitrate = bitrate
				bestAudioURL = format.URL
			}
		}
	}

	if bestAudioURL == "" {
		return nil, fmt.Errorf("没有找到音频流")
	}

	return &AudioInfo{
		URL:      bestAudioURL,
		Title:    videoInfo.Title,
		Author:   videoInfo.Author,
		Duration: formatDuration(videoInfo.LengthSeconds),
	}, nil
}

func formatDuration(seconds int) string {
	hours := seconds / 3600
	minutes := (seconds % 3600) / 60
	secs := seconds % 60

	if hours > 0 {
		return fmt.Sprintf("%d:%02d:%02d", hours, minutes, secs)
	}
	return fmt.Sprintf("%d:%02d", minutes, secs)
}

func extractVideoID(videoURL string) (string, error) {
	u, err := url.Parse(videoURL)
	if err != nil {
		return "", err
	}

	if u.Host == "youtu.be" {
		return u.Path[1:], nil
	}

	query := u.Query()
	videoID := query.Get("v")
	if videoID == "" {
		return "", fmt.Errorf("无法从URL中提取视频ID")
	}

	return videoID, nil
}
